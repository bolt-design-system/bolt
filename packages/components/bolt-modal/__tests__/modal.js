/* eslint-disable no-await-in-loop */
import {
  render,
  stopServer,
  html,
  vrtDefaultConfig as vrtConfig,
} from '../../../testing/testing-helpers';

const { readYamlFileSync } = require('@bolt/build-tools/utils/yaml');
const { join } = require('path');
const schema = readYamlFileSync(join(__dirname, '../modal.schema.yml'));
const { width, spacing, theme, scroll } = schema.properties;

const vrtDefaultConfig = Object.assign(vrtConfig, {
  failureThreshold: '0.02',
});

const timeout = 120000;

// Currently, the only important breakpoints to test are 'small' and 'large'
const viewportSizes = [
  {
    size: 'large',
    width: 1024,
    height: 768,
  },
  {
    size: 'small',
    width: 320,
    height: 568,
  },
];

describe('<bolt-modal> Component', () => {
  let page;

  beforeEach(async () => {
    await page.evaluate(() => {
      document.body.innerHTML = '';
    });
  }, timeout);

  beforeAll(async () => {
    page = await global.__BROWSER__.newPage();
    await page.goto('http://127.0.0.1:4444/', {
      timeout: 0,
    });
  }, timeout);

  afterAll(async () => {
    await stopServer();
    await page.close();
  }, timeout);

  test('basic usage', async () => {
    const results = await render('@bolt-components-modal/modal.twig', {
      content: 'This is a modal',
    });
    expect(results.ok).toBe(true);
    expect(results.html).toMatchSnapshot();
  });

  width.enum.forEach(async widthChoice => {
    test(`modal width: ${widthChoice}`, async () => {
      const results = await render('@bolt-components-modal/modal.twig', {
        width: widthChoice,
        content: 'This is a modal',
      });
      expect(results.ok).toBe(true);
      expect(results.html).toMatchSnapshot();
    });
  });

  spacing.enum.forEach(async spacingChoice => {
    test(`modal spacing: ${spacingChoice}`, async () => {
      const results = await render('@bolt-components-modal/modal.twig', {
        spacing: spacingChoice,
        content: 'This is a modal',
      });
      expect(results.ok).toBe(true);
      expect(results.html).toMatchSnapshot();
    });
  });

  theme.enum.forEach(async themeChoice => {
    test(`modal theme: ${themeChoice}`, async () => {
      const results = await render('@bolt-components-modal/modal.twig', {
        theme: themeChoice,
        content: 'This is a modal',
      });
      expect(results.ok).toBe(true);
      expect(results.html).toMatchSnapshot();
    });
  });

  scroll.enum.forEach(async scrollChoice => {
    test(`modal scroll: ${scrollChoice}`, async () => {
      const results = await render('@bolt-components-modal/modal.twig', {
        scroll: scrollChoice,
        content: 'This is a modal',
      });
      expect(results.ok).toBe(true);
      expect(results.html).toMatchSnapshot();
    });
  });

  test('Default <bolt-modal> with Shadow DOM renders', async function() {
    const renderedModal = await page.evaluate(async () => {
      const modal = document.createElement('bolt-modal');
      modal.setAttribute('uuid', '12345');
      modal.innerHTML = `<bolt-text tag="h3" slot="header">This is the header</bolt-text>
      <bolt-text>This is the body (default).</bolt-text>
      <bolt-text slot="footer">This is the footer</bolt-text>`;
      document.body.appendChild(modal);
      modal.updated();
      modal.show();
      return modal.outerHTML;
    });

    // const activeTagName = await page.evaluate(async () => {
    //   return document.activeElement.tagName;
    // });

    const renderedHTML = await html(renderedModal);

    await page.waitFor(1000); // wait a second before testing
    const image = await page.screenshot();

    // @todo: Fix this, returns 'BOLT-MODAL', expected 'BOLT-BUTTON'.
    // console.log(activeTagName);
    // expect(renderedModal.activeTagName === 'BOLT-BUTTON').toBe(true);

    expect(image).toMatchImageSnapshot(vrtDefaultConfig);

    expect(renderedHTML).toMatchSnapshot();
  });

  test('Default <bolt-modal> w/o Shadow DOM renders', async function() {
    const renderedModal = await page.evaluate(() => {
      const modal = document.createElement('bolt-modal');
      modal.setAttribute('uuid', '12345');
      modal.innerHTML = `<bolt-text tag="h3" slot="header">This is the header</bolt-text>
      <bolt-text>This is the body (default).</bolt-text>
      <bolt-text slot="footer">This is the footer</bolt-text>`;
      document.body.appendChild(modal);
      modal.useShadow = false;
      modal.updated();
      modal.show();
      return modal.outerHTML;
    });

    const renderedHTML = await html(renderedModal);

    await page.waitFor(1000); // wait a second before testing
    const image = await page.screenshot();

    expect(image).toMatchImageSnapshot(vrtDefaultConfig);

    expect(renderedHTML).toMatchSnapshot();
  });

  test(
    '<bolt-modal> rendered by Twig',
    async () => {
      const { html, ok } = await render('@bolt-components-modal/modal.twig', {
        content:
          '<bolt-text tag="h3" slot="header">This is the header</bolt-text><bolt-text>This is the body (default).</bolt-text><bolt-text slot="footer">This is the footer</bolt-text>',
      });
      expect(ok).toBe(true);
      expect(html).toMatchSnapshot();

      await page.evaluate(html => {
        const div = document.createElement('div');
        const trigger = document.createElement('bolt-button');
        trigger.textContent = 'Open Modal';
        trigger.setAttribute('class', 'js-modal-trigger--open');
        trigger.setAttribute('onclick', 'this.nextElementSibling.show()');
        div.innerHTML = `${html}`;
        div.prepend(trigger);
        document.body.appendChild(div);
      }, html);

      const screenshots = [];

      async function isVisible(selector) {
        return await page.evaluate(selector => {
          const e = document.querySelector(selector);
          if (!e) return false;
          const style = window.getComputedStyle(e);
          return style &&
            style.display !== 'none' &&
            style.visibility !== 'hidden' &&
            style.opacity !== '0'
            ? true
            : false;
        }, selector);
      }

      for (const item of viewportSizes) {
        const { height, width, size } = item;
        screenshots[size] = [];

        await page.setViewport({ height, width });
        await page.tap('.js-modal-trigger--open');
        await page.waitFor(500);

        if (await isVisible('bolt-modal')) {
          screenshots[size].modalOpened = await page.screenshot();
          expect(screenshots[size].modalOpened).toMatchImageSnapshot(
            vrtDefaultConfig,
          );
          await page.tap('bolt-button'); // closes modal
          await page.waitFor(500);
        }
      }
    },
    timeout,
  );
});
