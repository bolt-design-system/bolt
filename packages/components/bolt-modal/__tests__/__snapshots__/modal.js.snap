// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<bolt-modal> Component <bolt-modal> at various viewport sizes 1`] = `
<bolt-modal uuid="12345">
  <bolt-text tag="h3"
             slot="header"
  >
    This is the header
  </bolt-text>
  <bolt-text>
    This is the body (default).
  </bolt-text>
  <bolt-text slot="footer">
    This is the footer
  </bolt-text>
</bolt-modal>
`;

exports[`<bolt-modal> Component Default <bolt-modal> w/o Shadow DOM renders 2`] = `
<bolt-modal
  ready=""
  style=""
  uuid="12345"
>
  <!---->
  
      
      
  <div
    aria-hidden="true"
    class="c-bolt-modal c-bolt-modal--scroll-container c-bolt-modal--overlay-dark"
    style=""
  >
    
        
    <!---->
    <!---->
    
        
    <focus-trap
      style=""
    >
      
          
      <dialog
        aria-hidden="true"
        aria-labelledby="dialog-title-12345"
        class="c-bolt-modal__content c-bolt-modal__content--width-optimal"
        style=""
      >
        
            
        <article
          class="c-bolt-modal__container c-bolt-modal__container--spacing-medium t-bolt-xlight"
          style=""
        >
          
              
          <div
            class="c-bolt-modal__close-button c-bolt-modal__close-button--light"
            style=""
          >
            
                
      
            <bolt-trigger
              autofocus=""
              class="js-close-button-fallback"
              display="block"
              no-outline=""
              style=""
              tabindex="0"
            >
              
        
              <span
                class="c-bolt-modal__close-button__text"
                style=""
              >
                Close this dialog window
              </span>
              
        
              <span
                class="c-bolt-modal__close-button__icon"
                style=""
              />
              
      
            </bolt-trigger>
            
    
              
          </div>
          
              
          <header
            class="c-bolt-modal__container-header"
            style=""
          >
            
                
            <h1
              class="c-bolt-modal__dialog-title"
              id="dialog-title-12345"
              style=""
            >
              
                  Dialog content
                
            </h1>
            
                
            <!---->
            <!---->
            <bolt-text
              class="is-first-child is-last-child"
              slot="header"
              style=""
              tag="h3"
            >
              This is the header
            </bolt-text>
            <!---->
            <!---->
            
              
          </header>
          
              
          <div
            class="c-bolt-modal__container-body"
            style=""
          >
            
                
            <!---->
            <!---->
            
      
            <!---->
            <bolt-text
              class="is-first-child is-last-child"
              style=""
            >
              This is the body (default).
            </bolt-text>
            <!---->
            
      
            <!---->
            <!---->
            
              
          </div>
          
              
          
          <footer
            class="c-bolt-modal__container-footer"
            style=""
          >
            
            
            <!---->
            <!---->
            <bolt-text
              class="is-first-child is-last-child"
              slot="footer"
              style=""
            >
              This is the footer
            </bolt-text>
            <!---->
            <!---->
            
          
          </footer>
          
        
            
        </article>
        
          
      </dialog>
      
        
    </focus-trap>
    
      
  </div>
  
    
  <!---->
</bolt-modal>
`;

exports[`<bolt-modal> Component Default <bolt-modal> with Shadow DOM renders 2`] = `
<bolt-modal
  ready=""
  style=""
  uuid="12345"
>
  <bolt-text
    class="is-first-child is-last-child"
    slot="header"
    style=""
    tag="h3"
  >
    This is the header
  </bolt-text>
  
      
  <bolt-text
    class="is-first-child is-last-child"
    style=""
  >
    This is the body (default).
  </bolt-text>
  
      
  <bolt-text
    class="is-first-child is-last-child"
    slot="footer"
    style=""
  >
    This is the footer
  </bolt-text>
</bolt-modal>
`;

exports[`<bolt-modal> Component basic usage 1`] = `
<bolt-modal uuid="12345">
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal scroll: container 1`] = `
<bolt-modal scroll="container"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal scroll: overall 1`] = `
<bolt-modal scroll="overall"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal spacing: large 1`] = `
<bolt-modal spacing="large"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal spacing: medium 1`] = `
<bolt-modal spacing="medium"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal spacing: none 1`] = `
<bolt-modal spacing="none"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal spacing: small 1`] = `
<bolt-modal spacing="small"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal theme: dark 1`] = `
<bolt-modal theme="dark"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal theme: light 1`] = `
<bolt-modal theme="light"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal theme: none 1`] = `
<bolt-modal theme="none"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal theme: xdark 1`] = `
<bolt-modal theme="xdark"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal theme: xlight 1`] = `
<bolt-modal theme="xlight"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal width: auto 1`] = `
<bolt-modal width="auto"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal width: full 1`] = `
<bolt-modal width="full"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal width: optimal 1`] = `
<bolt-modal width="optimal"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;

exports[`<bolt-modal> Component modal width: regular 1`] = `
<bolt-modal width="regular"
            uuid="12345"
>
  This is a modal
</bolt-modal>
`;
